Tests/UnitTests/Purchasing/ProductsFetcherSK1Tests.swift

```diff
@@ -186,4 +186,21 @@ class ProductsFetcherSK1Tests: TestCase {
         expect(request.cancelCalled) == false
     }
 
+    func testMakesNewRequestAfterClearingCachedProductCorrectly() {
```

Comment:
Would it be possible to add a similar test for SK2's fetcher?

Reply:
Yeah, it could be great. TBH, I didn't do it because I don't know exactly to validate we fetch the products after clearing the cache. 

I think a possible way is:

```
1. Cache a product
2. Clear the cache
3. Fetch the products.
4. Validate that the obtained products are not the same as I previously cached.5. 
```

The problem is [the method to cache products is `private`](https://github.com/RevenueCat/purchases-ios/blob/588d00bdd8493003705057160b47df326bfc6ebc/Sources/Purchasing/StoreKit2/ProductsFetcherSK2.swift#L69), and it's not available for testing. We could make it internal just for testing purposes like [we did for `ProductsFetcherSK1`](https://github.com/RevenueCat/purchases-ios/blob/588d00bdd8493003705057160b47df326bfc6ebc/Sources/Purchasing/StoreKit1/ProductsFetcherSK1.swift#L185)...WDYT? 🤔

Reply:
Hmm yeah I don't see an easy way without exposing implementation details, so never mind.

Reply:
I suppose our choices at this point are: 
1. Exposing the fact that `ProductsFetcherSK2` caches objects at an `internal` level, solely for the purpose of testing
2. Not testing the caching that does happen in `ProductsFetcherSK2`
3. A larger refactor to introduce DI, but I can't think of a way of doing that that doesn't conceptually do the same as 1.

I'd opt for 1., I'd rather have a bit more test coverage than be overly purist with the internal API here

Reply:
Yep, we could do it the option 1. In fact, we are doing the same for the `ProductsFetcherSK1` 🙂

> in tests, would we be able to do an extension ProductsFetcherSK2 that introduces a variable that exposes the cached products? It's still conceptually the same as 1., just doing it in the Unit tests code instead of the main implementation.

As far as I know, it's not possible. We only can use the `public` code for testing, and`internal` code is possible if we include the framework as `@testable`, but `private` it's not possible.